### PR TITLE
Flexible styling of attributed string renderer

### PR DESCRIPTION
**Pull request by jlj**
This PR is about bringing flexibility and ease-of-use to the attributed string renderer customization API.

Example of customization:

````swift
let textAttributes = CMTextAttributes()!;

// Customize the color and font of header elements
textAttributes.addStringAttributes([ .foregroundColor: UIColor(red: 0.0, green: 0.446, blue: 0.657, alpha: 1.0)], 
                                   forElementWithKinds: .anyHeader)
let boldItalicTrait: UIFontDescriptor.SymbolicTraits = [.traitBold, .traitItalic];
textAttributes.addFontAttributes([ .family: "Avenir Next" ,
                                   .traits: [ UIFontDescriptor.TraitKey.symbolic: boldItalicTrait.rawValue]], 
                                 forElementWithKinds: .anyHeader)
textAttributes.setFontTraits([.weight: UIFont.Weight.heavy], forElementWithKinds: [.header1, .header2])
textAttributes.addFontAttributes([ .size: 40 ], forElementWithKinds: .header1)

// Customize the font and paragraph alignment of block-quotes
textAttributes.addFontAttributes([.family: "Noteworthy", .size: 18], forElementWithKinds: .blockQuote)
textAttributes.addParagraphStyleAttributes([ .alignment: NSTextAlignment.center.rawValue], forElementWithKinds: .blockQuote)

// Customize the background color of code elements
textAttributes.addStringAttributes([ .backgroundColor: UIColor(white: 0.9, alpha: 0.5)], 
                                   forElementWithKinds: [.inlineCode, .codeBlock])

let renderer = CMAttributedStringRenderer(document: document, attributes: textAttributes)!
````

renders as:

Capture d’écran 2019-11-04 à 13 55 46

Note that on iOS, the rendered attributed string is fully compliant with dynamic type when customized font attributes are used.

Internal details (commit message):
- New class `CMStyleAttributes` provides an easier more-flexible way of styling Markdown elements. By exposing font attributes and paragraph attributes separately from other string attributes, `CMStyleAttributes` provides a simple way of  cascading attributes like font traits or paragraph headings.
- New incremental attributes setting APIs in `CMTextAttributes` allowing to add attributes to multiple Markdown element types at once.
- Attributes for element types strong and emphasis are now defined as regular defaults using font attributes.
- Simplified `CMCascadingAttributeStack` push/pop api.
- Updated README with description of the new attributes setting API.
- Updated example apps by adding attributes configuration sample code.

Misc:
- Render a code block as a single TextKit paragraph by replacing internal '\n' with line separators. This makes the definition of paragraph attributes on code blocks element consistent with other block types.

### CODE REVIEW

File: Example-Mac/AppDelegate.swift [hunk 1]
```diff
@@ -55,7 +55,28 @@ class AppDelegate: NSObject, NSApplicationDelegate {
             document?.linksBaseURL = linksBaseUrl
         }
         
-        let renderer = CMAttributedStringRenderer(document: document, attributes: CMTextAttributes())!
+        let textAttributes = CMTextAttributes()!;
```
Comment (indragiek) [inline Example-Mac/AppDelegate.swift]: Is the force unwrap for CMTextAttributes necessary here? It looks like -[CMTextAttributes init] never returns nil, so putting a `nonnull` annotation on the return type might help avoid this.

File: Example-iOS/ViewController.swift [hunk 2]
```diff
@@ -17,7 +17,25 @@ class ViewController: UIViewController {
         let path = Bundle.main.path(forResource: "test", ofType: "md")!
         let document = CMDocument(contentsOfFile: path, options: CMDocumentOptions(rawValue: 0))
         
-        let renderer = CMAttributedStringRenderer(document: document, attributes: CMTextAttributes())!
+        let textAttributes = CMTextAttributes()!;
```
Comment (indragiek) [inline Example-iOS/ViewController.swift]: here too

File: Example-Mac/AppDelegate.swift (hunk 1 above)
Comment (indragiek) [reply Example-Mac/AppDelegate.swift]: Also, remove semicolon at end of line

Comment (jlj) [reply Example-Mac/AppDelegate.swift]: Correct. Fixed by assuming non-null in `CMTextAttributes` in commit a5db7ab81f364a6605b601c1a8bf222fa9fb4322

File: Example-iOS/ViewController.swift (hunk 2 above)
Comment (jlj) [reply Example-iOS/ViewController.swift]: Also done in commit a5db7ab81f364a6605b601c1a8bf222fa9fb4322

**Comment by jlj** [comment]
For the record, a few possible improvement ideas that we could add later:
- better list items alignment regarding continuation lines and some configurability of bullets and item numbers formats (typically as paragraph attributes) ,
- caching of cascaded attributes at every level to avoid recomputing the whole stack after every push or pop.

**Comment by indragiek** [comment]
Thank you again!

**Pull request merged**